### ISSUE TITLE
Cors headers on Alpha4 not return Allow Origin

### ISSUE THREAD

**Issue by felpasl** [containous/traefik#4825]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [X] No

### What did you expect to see?

Cors configuration Working

### What did you see instead?

Access-Control-Allow-Origin header with Origin value 

### Output of `traefik version`: (_What version of Traefik are you using?_)

I compiled this image to debug logs, trying further investigate the issue it's based on the pull of yesterday master (alpha4)

```
Version:      11f04a453ef9a255c4b6bd3cff7e1938ee3e16ba
Codename:     cheddar
Go version:   go1.12.4
Built:        2019-04-25_09:07:21PM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml

[global]
  checkNewVersion = true
  sendAnonymousUsage = true
  debug = true

[entryPoints]
    [entryPoints.web]
    address = ":80"

[log]

level = "DEBUG"

[accessLog]

[providers.docker]

exposedByDefault = false
#swarmMode = true
```

```
version: '3.7'

services:
  reverse-proxy:
    image: traefiklocal # The official v2.0 Traefik docker image
    #command: --api --providers.docker # Enables the web UI and tells Traefik to listen to docker
    ports:
      - "80:80"     # The HTTP port
      - "8080:8080" # The Web UI (enabled by --api)
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock # So that Traefik can listen to the Docker events# ...
      - ./traefik.toml:/etc/traefik/traefik.toml # So that Traefik can listen to the Docker events
    networks:
      web:
  whoami:
    image: containous/whoami # A container that exposes an API to show its IP address
    #deploy:
    labels:
      - "traefik.http.routers.whoami.entrypoints=web"
      - "traefik.http.routers.whoami.rule=Host(`whoami.docker.localhost`)"
      - "traefik.http.routers.whoami.service=whoami" # <-----
      - "traefik.http.routers.whoami.middlewares=testHeader" # <-----
      - "traefik.http.services.whoami.loadbalancer.server.port=80"
      - "traefik.http.middlewares.testHeader.Headers.AccessControlAllowMethods=GET,OPTIONS,PUT,DELETE"
      - "traefik.http.middlewares.testHeader.Headers.AccessControlAllowOrigin=origin-list-or-null"
      - "traefik.http.middlewares.testHeader.Headers.AccessControlMaxAge=100"
      - "traefik.enable=true"
      - "traefik.port=80"
    networks:
      web:

networks:
  web:
```

### Curl Requests:
**Response OPTIONS OK**
```
curl -X OPTIONS   http://whoami.docker.localhost   -H 'Access-Control-Request- Headers: Origin'   -H 'Access-Control-Request-Method: GET'   -H 'Origin: http:// localhost/'   -H 'Postman-Token: 6e6b862a-4dae-40f6-927f-55d78ce31e00'   -H 'cac he-control: no-cache' -i

HT TP/1.1 200 OK                                                                    
Access-Control-Allow-Methods: GET,OPTIONS,PUT,DELETE                             
Access-Control-Allow-Origin: http://localhost/                                   
Access-Control-Max-Age: 100                                                      
Date: Fri, 26 Apr 2019 13:03:11 GMT                                              
Content-Length: 0
--
```
**Actual Response Cors**
Access-Control-Allow-Origin: **null**  
**Expected Response**:
Access-Control-Allow-Origin: **http://localhost/**      
```
curl -X GET http://whoami.docker.localhost -H 'Origin: http://localhost/' -H 'Postman-Token: ffc2a96f-16ba-4266-b8f4-19d15fd3b438' -H 'cache-control: no-cache' -i

HT TP/1.1 200 OK                                                                    
Access-Control-Allow-Origin: null                                                
Content-Length: 447                                                              
Content-Type: text/plain; charset=utf-8                                          
Date: Fri, 26 Apr 2019 13:05:26 GMT                                              
X-Request-Origin: http://localhost/                                                                                                                               


Hostname: 987f738fbb0e                                                           
IP: 127.0.0.1                                                                    
IP: 172.19.0.2                                                                   
GET / HTTP/1.1                                                                   
Host: whoami.docker.localhost                                                    
User-Agent: curl/7.63.0                                                          
Accept: */*                                                                      
Accept-Encoding: gzip                                                            
Cache-Control: no-cache                                                          
Origin: http://localhost/                                                        
Postman-Token: ffc2a96f-16ba-4266-b8f4-19d15fd3b438                              
X-Forwarded-For: 192.168.56.1                                                    
X-Forwarded-Host: whoami.docker.localhost                                        
X-Forwarded-Port: 80                                                             
X-Forwarded-Proto: http                                                         
X-Forwarded-Server: 031bcb4dc4eb                                                 
X-Real-Ip: 192.168.56.1
```

Another Question: Is it in the plan to create a config to list which origins are enabled to have Cors returns?

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="ServeHTTP originHeaderhttp://localhost/" middlewareName=docker.testHeader entryPointName=web middlewareType=Headers routerName=docker.whoami
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="vulcand/oxy/roundrobin/rr: begin ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Origin\":[\"http://localhost/\"],\"Postman-Token\":[\"59567cd7-d216-4a02-8dc3-764885d65260\"],\"User-Agent\":[\"PostmanRuntime/7.6.0\"],\"X-Forwarded-Host\":[\"whoami.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"031bcb4dc4eb\"],\"X-Real-Ip\":[\"192.168.56.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"whoami.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"192.168.56.1:52082\",\"RequestURI\":\"/\",\"TLS\":null}"
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="vulcand/oxy/roundrobin/rr: Forwarding this request to URL" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Origin\":[\"http://localhost/\"],\"Postman-Token\":[\"59567cd7-d216-4a02-8dc3-764885d65260\"],\"User-Agent\":[\"PostmanRuntime/7.6.0\"],\"X-Forwarded-Host\":[\"whoami.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"031bcb4dc4eb\"],\"X-Real-Ip\":[\"192.168.56.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"whoami.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"192.168.56.1:52082\",\"RequestURI\":\"/\",\"TLS\":null}" ForwardURL="http://172.19.0.2:80"
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg=ModifyResponseHeaders routerName=docker.whoami middlewareName=docker.testHeader entryPointName=web middlewareType=Headers
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="ModifyResponseHeaders originHeaderKey:X-Request-Origin" entryPointName=web middlewareType=Headers routerName=docker.whoami middlewareName=docker.testHeader
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="ModifyResponseHeaders originHeader:" routerName=docker.whoami middlewareName=docker.testHeader entryPointName=web middlewareType=Headers
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="ModifyResponseHeaders allowOrigin:null" routerName=docker.whoami middlewareName=docker.testHeader entryPointName=web middlewareType=Headers
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="ModifyResponseHeaders res.Header.Del(originHeaderKey):X-Request-Origin" entryPointName=web middlewareType=Headers routerName=docker.whoami middlewareName=docker.testHeader
reverse-proxy_1  | time="2019-04-26T12:53:06Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"GET\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"*/*\"],\"Accept-Encoding\":[\"gzip, deflate\"],\"Cache-Control\":[\"no-cache\"],\"Connection\":[\"keep-alive\"],\"Origin\":[\"http://localhost/\"],\"Postman-Token\":[\"59567cd7-d216-4a02-8dc3-764885d65260\"],\"User-Agent\":[\"PostmanRuntime/7.6.0\"],\"X-Forwarded-Host\":[\"whoami.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"031bcb4dc4eb\"],\"X-Real-Ip\":[\"192.168.56.1\"]},\"ContentLength\":0,\"TransferEncoding\":null,\"Host\":\"whoami.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"192.168.56.1:52082\",\"RequestURI\":\"/\",\"TLS\":null}"
reverse-proxy_1  | 192.168.56.1 - - [26/Apr/2019:12:53:06 +0000] "GET / HTTP/1.1" 200 465 "-" "PostmanRuntime/7.6.0" 4 "docker.whoami" "http://172.19.0.2:80" 3ms
```

this is the header.go file modified to generate get this log:
```Go
// Package headers Middleware based on https://github.com/unrolled/secure.
package headers

import (
	"context"
	"errors"
	"net/http"
	"strconv"
	"strings"

	"github.com/containous/traefik/pkg/config"
	"github.com/containous/traefik/pkg/middlewares"
	"github.com/containous/traefik/pkg/tracing"
	"github.com/opentracing/opentracing-go/ext"
	"github.com/unrolled/secure"
	"github.com/sirupsen/logrus"
)

const (
	typeName        = "Headers"
	originHeaderKey = "X-Request-Origin"
)

type headers struct {
	name    string
	handler http.Handler
}
var logger logrus.FieldLogger

// New creates a Headers middleware.
func New(ctx context.Context, next http.Handler, config config.Headers, name string) (http.Handler, error) {
	// HeaderMiddleware -> SecureMiddleWare -> next
	logger = middlewares.GetLogger(ctx, name, typeName)
	logger.Debug("Creating middleware")

	hasSecureHeaders := config.HasSecureHeadersDefined()
	hasCustomHeaders := config.HasCustomHeadersDefined()
	hasCorsHeaders := config.HasCorsHeadersDefined()

	if !hasSecureHeaders && !hasCustomHeaders && !hasCorsHeaders {
		return nil, errors.New("headers configuration not valid")
	}

	var handler http.Handler
	nextHandler := next

	if hasSecureHeaders {
		logger.Debug("Setting up secureHeaders from %v", config)
		handler = newSecure(next, config)
		nextHandler = handler
	}

	if hasCustomHeaders || hasCorsHeaders {
		logger.Debug("Setting up customHeaders/Cors from %v", config)
		handler = NewHeader(nextHandler, config)
	}

	return &headers{
		handler: handler,
		name:    name,
	}, nil
}

func (h *headers) GetTracingInformation() (string, ext.SpanKindEnum) {
	return h.name, tracing.SpanKindNoneEnum
}

func (h *headers) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
	h.handler.ServeHTTP(rw, req)
}

type secureHeader struct {
	next   http.Handler
	secure *secure.Secure
}

// newSecure constructs a new secure instance with supplied options.
func newSecure(next http.Handler, headers config.Headers) *secureHeader {
	opt := secure.Options{
		BrowserXssFilter:        headers.BrowserXSSFilter,
		ContentTypeNosniff:      headers.ContentTypeNosniff,
		ForceSTSHeader:          headers.ForceSTSHeader,
		FrameDeny:               headers.FrameDeny,
		IsDevelopment:           headers.IsDevelopment,
		SSLRedirect:             headers.SSLRedirect,
		SSLForceHost:            headers.SSLForceHost,
		SSLTemporaryRedirect:    headers.SSLTemporaryRedirect,
		STSIncludeSubdomains:    headers.STSIncludeSubdomains,
		STSPreload:              headers.STSPreload,
		ContentSecurityPolicy:   headers.ContentSecurityPolicy,
		CustomBrowserXssValue:   headers.CustomBrowserXSSValue,
		CustomFrameOptionsValue: headers.CustomFrameOptionsValue,
		PublicKey:               headers.PublicKey,
		ReferrerPolicy:          headers.ReferrerPolicy,
		SSLHost:                 headers.SSLHost,
		AllowedHosts:            headers.AllowedHosts,
		HostsProxyHeaders:       headers.HostsProxyHeaders,
		SSLProxyHeaders:         headers.SSLProxyHeaders,
		STSSeconds:              headers.STSSeconds,
	}

	return &secureHeader{
		next:   next,
		secure: secure.New(opt),
	}
}

func (s secureHeader) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
	s.secure.HandlerFuncWithNextForRequestOnly(rw, req, s.next.ServeHTTP)
}

// Header is a middleware that helps setup a few basic security features. A single headerOptions struct can be
// provided to configure which features should be enabled, and the ability to override a few of the default values.
type Header struct {
	next    http.Handler
	headers *config.Headers
}

// NewHeader constructs a new header instance from supplied frontend header struct.
func NewHeader(next http.Handler, headers config.Headers) *Header {
	return &Header{
		next:    next,
		headers: &headers,
	}
}

func (s *Header) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
	reqAcMethod := req.Header.Get("Access-Control-Request-Method")
	reqAcHeaders := req.Header.Get("Access-Control-Request-Headers")
	originHeader := req.Header.Get("Origin")

	logger.Debug("ServeHTTP originHeader", originHeader)

	if reqAcMethod != "" && reqAcHeaders != "" && originHeader != "" && req.Method == http.MethodOptions {
		// If the request is an OPTIONS request with an Access-Control-Request-Method header, and  Access-Control-Request-Headers headers,
		// and Origin headers, then it is a CORS preflight request, and we need to build a custom response: https://www.w3.org/TR/cors/#preflight-request
		if s.headers.AccessControlAllowCredentials {
			rw.Header().Set("Access-Control-Allow-Credentials", "true")
		}

		allowHeaders := strings.Join(s.headers.AccessControlAllowHeaders, ",")
		if allowHeaders != "" {
			rw.Header().Set("Access-Control-Allow-Headers", allowHeaders)
		}

		allowMethods := strings.Join(s.headers.AccessControlAllowMethods, ",")
		if allowMethods != "" {
			rw.Header().Set("Access-Control-Allow-Methods", allowMethods)
		}

		allowOrigin := s.getAllowOrigin(originHeader)

		if allowOrigin != "" {
			rw.Header().Set("Access-Control-Allow-Origin", allowOrigin)
		}

		logger.Debug("ServeHTTP allowOrigin", allowOrigin)

		rw.Header().Set("Access-Control-Max-Age", strconv.Itoa(int(s.headers.AccessControlMaxAge)))

		return
	}

	if len(originHeader) > 0 {
		rw.Header().Set(originHeaderKey, originHeader)
	}

	s.modifyRequestHeaders(req)
	// If there is a next, call it.
	if s.next != nil {
		s.next.ServeHTTP(rw, req)
	}
}

// modifyRequestHeaders sets or deletes request headers.
func (s *Header) modifyRequestHeaders(req *http.Request) {
	// Loop through Custom request headers
	for header, value := range s.headers.CustomRequestHeaders {
		if value == "" {
			req.Header.Del(header)
		} else {
			req.Header.Set(header, value)
		}
	}
}

// ModifyResponseHeaders set or delete response headers
func (s *Header) ModifyResponseHeaders(res *http.Response) error {
	
	logger.Debug("ModifyResponseHeaders")

	// Loop through Custom response headers
	for header, value := range s.headers.CustomResponseHeaders {
		if value == "" {
			res.Header.Del(header)
		} else {
			res.Header.Set(header, value)
		}
	}
	logger.Debug("ModifyResponseHeaders originHeaderKey:", originHeaderKey)

	originHeader := res.Header.Get(originHeaderKey)
	
	logger.Debug("ModifyResponseHeaders originHeader:", originHeader)

	allowOrigin := s.getAllowOrigin(originHeader)
	
	logger.Debug("ModifyResponseHeaders allowOrigin:", allowOrigin)
	
	//	Delete the origin header key, since it is only used to pass data from the request for response handling
	res.Header.Del(originHeaderKey)
	
	logger.Debug("ModifyResponseHeaders res.Header.Del(originHeaderKey):", originHeaderKey)

	if allowOrigin != "" {
		res.Header.Set("Access-Control-Allow-Origin", allowOrigin)

		if s.headers.AddVaryHeader {
			varyHeader := res.Header.Get("Vary")
			if varyHeader != "" {
				varyHeader += ","
			}
			varyHeader += "Origin"

			res.Header.Set("Vary", varyHeader)
		}
	}

	if s.headers.AccessControlAllowCredentials {
		res.Header.Set("Access-Control-Allow-Credentials", "true")
	}

	exposeHeaders := strings.Join(s.headers.AccessControlExposeHeaders, ",")
	if exposeHeaders != "" {
		res.Header.Set("Access-Control-Expose-Headers", exposeHeaders)
	}

	return nil
}

func (s *Header) getAllowOrigin(header string) string {
	switch s.headers.AccessControlAllowOrigin {
	case "origin-list-or-null":
		if len(header) == 0 {
			return "null"
		}
		return header
	case "*":
		return "*"
	default:
		return ""
	}
}
```

**Comment by dtomcej** [comment]
Thanks for this!

@felpasl, does the exact same behavior exist in the official alpha 4 image?

I will be looking into this shortly!

**Comment by felpasl** [comment]
yes, this is the same behavior, I'm building the image to improve log about this, to view if I'm doing wrong config.

In another point, about this question: Is it in the plan to create a config to list which origins are enabled to have Cors returns?

**Comment by dtomcej** [comment]
@felpasl I have been able to reproduce your issue, and am currently working on a fix.

Thanks!

**Comment by traefiker** [comment]
Closed by #4857.
